mesh/silvair-io.c

```diff
@@ -132,13 +133,39 @@ struct silvair_keep_alive_cmd_pld {
 } __packed;
 
 
+static bool buffered_write(struct l_io *l_io, void *user_data)
+{
+    struct silvair_io *io = user_data;
+
+    (void)l_ringbuf_write(io->l_out_ringbuf, l_io_get_fd(l_io));
+
+    return (l_ringbuf_len(io->l_out_ringbuf) != 0);
+}
+
 static bool simple_write(struct silvair_io *io,
 				uint8_t *buf,
 				size_t size)
 {
-	int w = write(l_io_get_fd(io->l_io), buf, size);
+    int consumed;
+    ssize_t buffered;
+
+    consumed = write(l_io_get_fd(io->l_io), buf, size);
 
-	return (w > 0 && (size_t)w == size);
+    if ((size_t)consumed == size)
+    {
+        /* All data have been sent. */
+        return true;
+    }
+
+    buffered = l_ringbuf_append(io->l_out_ringbuf, buf, size);
+
+    if (buffered < 0)
+    {
+        /* There is no space in the ringbuf, all data have been dropped. */
```

Comment:
If you take a closer look at the way `io_write` is used, you'll notice that in kernel SLIP mode, it either sends or drops the whole protocol frame.

It would be nice if we could achieve the same with userspace SLIP.